### ISSUE TITLE
jax.numpy.arcsin not has accurate for complex64 dtype

### ISSUE THREAD

**Issue by samthakur587** [google/jax#19398]
### Description

```python
import numpy as np
import jax.numpy as jnp
from jax import jit
import torch
import paddle
import warnings

# Suppress warnings
warnings.filterwarnings("ignore")
# NumPy
complex_number_numpy = np.array([-1.0 - 101999.0j], dtype=np.complex64)
result_numpy = np.arcsinh(complex_number_numpy)
print("-------------------------------------------------------------------------------------------------------------------------------")
print("NumPy Result:", result_numpy)

import tensorflow as tf

# Define the complex number
complex_number = tf.constant([-1.0 - 101999.0j], dtype=tf.complex64)

# Calculate the inverse hyperbolic sine (asinh)
result = tf.math.asinh(complex_number)
print("---------------------------------------------------------------------------------------------------------------------------------------")
# Print the result
print("Tensorflow Result:", result.numpy())
# JAX
complex_number_jax = jnp.array([-1.0 - 101999.0j], dtype=jnp.complex64)
asinh_fn = jit(jnp.arcsinh)
result_jax = asinh_fn(complex_number_jax)
print("---------------------------------------------------------------------------------------------------------------------------------------")
print("JAX Result:", result_jax)

# PyTorch
complex_number_torch = torch.tensor([-1.0 - 101999.0j], dtype=torch.complex64)
result_torch = torch.asinh(complex_number_torch)
print("---------------------------------------------------------------------------------------------------------------------------------------")
print("PyTorch Result:", result_torch.numpy())

```

# Output
```text
-------------------------------------------------------------------------------------------------------------------------------
NumPy Result: [-12.225865-1.5707865j]
---------------------------------------------------------------------------------------------------------------------------------------
Tensorflow Result: [-12.225865-1.5707865j]
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1705497661.625412   30592 tfrt_cpu_pjrt_client.cc:349] TfrtCpuClient created.
[2024-01-17 18:51:01,626] [ WARNING] xla_bridge.py:527 - No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
---------------------------------------------------------------------------------------------------------------------------------------
JAX Result: [-5.348053+0.j]
---------------------------------------------------------------------------------------------------------------------------------------
PyTorch Result: [-12.225865-1.5707865j]
```

### What jax/jaxlib version are you using?

latest

### Which accelerator(s) are you using?

cpu

### Additional system info?

python 3.10

**Comment by jakevdp** [comment]
Thanks for the report. It looks like this is related to overflow of the complex type within the implementation; it's the issue mentioned in this TODO: https://github.com/openxla/xla/blob/64c6deaf73a80b5df90bd1e3c3df9ef5211fe976/xla/client/lib/math.cc#L1268-L1270

We'll probably need to move this bug upstream into OpenXLA.

**Comment by samthakur587** [comment]
Hii @jakevdp i have seen the issue of overflow. and i made i issue at XLA also . So we should  wait for the solution from XLA
or do you have any alternate solution to solve this problem in jax.

thanks

**Comment by jakevdp** [comment]
Cross reference: https://github.com/openxla/xla/issues/8553

**Comment by pearu** [comment]
The co-culprit appears to be jax sqrt that inaccuracy is notable in the region reported here:
```python
>>> import numpy as np
>>> import jax.numpy as jnp
>>> x = -1 - 100000j
>>> np.sqrt(np.array(x**2 + 1, dtype=np.complex64))       # numpy sqrt
(1+100000j)
>>> jnp.sqrt(jnp.array(x**2 + 1, dtype=jnp.complex64))    # jax sqrt
Array(0.9969869+100000.j, dtype=complex64)
>>> jnp.sqrt(jnp.array(x**2 + 1, dtype=jnp.complex128))   # reference result
Array(1.+99999.999995j, dtype=complex128)
```
Notice that the numpy sqrt result demonstrates overflow but jax sqrt result is just inaccurate.

That said, improving jax sqrt accuracy will not resolve this issue because of the overflow.
To avoid the overflow, we could use the identity `x**2 + 1 == (x + 1j) * (x - 1j)` that will lead to a better approximation of jax arcsinh so that its result will match numpy arcsinh result:
```python
>>> jnp.log(jnp.sqrt(jnp.array(x + 1j, dtype=jnp.complex64)) * jnp.sqrt(jnp.array(x - 1j, dtype=jnp.complex64)) + x)
Array(12.206073-1.5708064j, dtype=complex64)
>>> np.log(np.sqrt(np.array(x + 1j, dtype=np.complex64)) * np.sqrt(np.array(x - 1j, dtype=np.complex64)) + x)
(12.206072645580173-1.5708063267948962j)
```

**Comment by pearu** [comment]
Btw, the source code for asinh on complex inputs is
https://github.com/openxla/xla/blob/90a45af8ae9dc2c9a7aba20d8bf8c31b8e9fab65/xla/mlir_hlo/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo_patterns.td#L260

**Comment by pearu** [comment]
I have submitted a fix as https://github.com/openxla/xla/pull/8589.
With that, we have:
```
NumPy Result: [-12.225865-1.5707865j]
JAX Result: [-12.2332945-1.5707966j]
```

**Comment by samthakur587** [comment]
Hii @pearu Thanks for the grate explanation. i have done lots of experiment based on your suggestion.

# Method 1
```python
>>> jnp.log(jnp.sqrt(jnp.array(x + 1j, dtype=jnp.complex64)) * jnp.sqrt(jnp.array(x - 1j, dtype=jnp.complex64)) + x)
Array(12.206073-1.5708064j, dtype=complex64)
>>> np.log(np.sqrt(np.array(x + 1j, dtype=np.complex64)) * np.sqrt(np.array(x - 1j, dtype=np.complex64)) + x)
(12.206072645580173-1.5708063267948962j)
```
This method fails on most of the tests even on small number.

# Method 2
```python
>>> np.arcsinh(x)
(-12.206072645555174-1.5707863267948965j)
```
This method is passed most of the cases but this also has low precise output. if it's imag value has larger then 1e6-7.

**Comment by samthakur587** [comment]
And only Asinh has not this problem even Asin and Acos and Acosh all of that has the same problem.

**Comment by samthakur587** [comment]
The method fails on. `-3.-3498384.j` this example.

**Comment by pearu** [comment]
@samthakur587 thanks for testing! 

A couple of notes:

Ideally, all frameworks ought to result the same value that is close to the reference value, say, the result computed on complex128 arguments. Btw, could you include such a reference value computation in your script?

There is an improvement of using Method 2 compared to the current state where overflow produces garbage, but it is not ideal. Other frameworks rely on C++ std implementations but XLA provides its own implementation for many elementary/special functions. So, to achieve the ideal, the algorithms used by C++ std and XLA should basically be the same. Atm, I am not sure which algorithm is used in C++ std asinh.

I think we can reject Method 1 (because it uses inaccurate sqrt) but I am still curious why it picks up a different branch (notice the sign difference in the real part).

`allclose` does not detect if the a different branch value of asinh is returned that otherwise is correct. The correctness of the result value can be tested via `|sinh(asinh(x)) - x|`. Although, ideally, all frameworks ought to use the same branch.

**Comment by pearu** [comment]
Btw, boost uses Method 2: https://www.boost.org/doc/libs/1_50_0/boost/math/complex/asinh.hpp

**Comment by samthakur587** [comment]
> Btw, boost uses Method 2: https://www.boost.org/doc/libs/1_50_0/boost/math/complex/asinh.hpp However, there are differences in XLA arcsin and boost https://www.boost.org/doc/libs/1_50_0/boost/math/complex/asin.hpp, hence the slightly different results reported above (it is likely that other frameworks use the same algorithms as boost).

I think first we have to correct the `asin` implementation. it's also not accurate for complex dtype. and has the same problem like `asinh`. and in method2 we are using the `asin` which is itself has not correct implementation.

**Comment by samthakur587** [comment]
Here's the in `Asin` implementation it also uses the `2 * atan(x / (1 + sqrt(1 - x^2)))` so it also has the `sqrt` that has overflow problem. https://github.com/openxla/xla/blob/90a45af8ae9dc2c9a7aba20d8bf8c31b8e9fab65/xla/mlir_hlo/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo_patterns.td#L147C1-L147C1

**Comment by rajasekharporeddy** [comment]
Hi @samthakur587

It looks like this issue was resolved with https://github.com/openxla/stablehlo/pull/2402. I tested the mentioned code with JAX version 0.4.31 and JAX's `jax.numpy.arcsinh` now produces the results consistent with NumPy, TensorFlow and PyTorch.

```python
import numpy as np
import jax.numpy as jnp
import tensorflow as tf
from jax import jit
import torch
import warnings

# Suppress warnings
warnings.filterwarnings("ignore")
# NumPy
complex_number_numpy = np.array([-1.0 - 101999.0j], dtype=np.complex64)
result_numpy = np.arcsinh(complex_number_numpy)
print("-------------------------------------------------------------------------------------------------------------------------------")
print("NumPy Result:", result_numpy)

# Define the complex number
complex_number = tf.constant([-1.0 - 101999.0j], dtype=tf.complex64)

# Calculate the inverse hyperbolic sine (asinh)
result = tf.math.asinh(complex_number)
print("---------------------------------------------------------------------------------------------------------------------------------------")
# Print the result
print("Tensorflow Result:", result.numpy())
# JAX
complex_number_jax = jnp.array([-1.0 - 101999.0j], dtype=jnp.complex64)
asinh_fn = jit(jnp.arcsinh)
result_jax = asinh_fn(complex_number_jax)
print("---------------------------------------------------------------------------------------------------------------------------------------")
print("JAX Result:", result_jax)

# PyTorch
complex_number_torch = torch.tensor([-1.0 - 101999.0j], dtype=torch.complex64)
result_torch = torch.asinh(complex_number_torch)
print("---------------------------------------------------------------------------------------------------------------------------------------")
print("PyTorch Result:", result_torch.numpy())
```
Output:
```
-------------------------------------------------------------------------------------------------------------------------------
NumPy Result: [-12.225865-1.5707865j]
---------------------------------------------------------------------------------------------------------------------------------------
Tensorflow Result: [-12.225865-1.5707865j]
---------------------------------------------------------------------------------------------------------------------------------------
JAX Result: [-12.225865-1.5707865j]
---------------------------------------------------------------------------------------------------------------------------------------
PyTorch Result: [-12.225865-1.5707865j]
```

```python
import jax.numpy as jnp
import numpy as np
import tensorflow as tf
import torch

# Ignore warnings
import warnings
warnings.filterwarnings("ignore")

def generate_complex_number():
    real_part = np.random.uniform(-10, 10)
    imag_part = np.random.uniform(-100000000, 100000000)
    return np.complex64(real_part + 1j * imag_part)

def main():
    for i in range(10):
        # Generate a random complex number
        x = generate_complex_number()
        print(f"Generated complex number: {x}")

        # NumPy calculation
        np_asinh = np.arcsinh(np.array(x, dtype=np.complex64))
        print(f"NumPy asinh result: {np_asinh}")

        # TensorFlow calculation
        tf_asinh = tf.math.asinh(tf.constant(x, dtype=tf.complex64))
        print(f"TensorFlow asinh result: {tf_asinh.numpy()}")

        # PyTorch calculation
        torch_asinh = torch.asinh(torch.tensor(x, dtype=torch.complex64))
        print(f"PyTorch asinh result: {torch_asinh.numpy()}")

        # JAX calculation - Method 1
        jax_asinh = jit(jnp.arcsinh)(jnp.array(x, dtype=jnp.complex64))
        print(f"JAX asinh result: {jax_asinh}")

        # Check Method 1 vs other frameworks
        np_vs_jax = np.allclose(np_asinh, jax_asinh, rtol=1e-2, atol=1e-2)
        tf_vs_jax = np.allclose(tf_asinh.numpy(), jax_asinh, rtol=1e-2, atol=1e-2)
        torch_vs_jax = np.allclose(torch_asinh.numpy(), jax_asinh, rtol=1e-2, atol=1e-2)

        print(f"NumPy vs JAX: {np_vs_jax}")
        print(f"TensorFlow vs JAX: {tf_vs_jax}")
        print(f"PyTorch vs JAX: {torch_vs_jax}")

        print("=" * 30)  # Separator for better output readability

if __name__ == "__main__":
    main()
```
Output:
```
Generated complex number: (3.6044483184814453+55384400j)
NumPy asinh result: (18.52295684814453+1.570796251296997j)
TensorFlow asinh result: (18.52295684814453+1.570796251296997j)
PyTorch asinh result: (18.52295684814453+1.570796251296997j)
JAX asinh result: (18.5229549407959+1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (5.806606769561768-95168112j)
NumPy asinh result: (19.064302444458008-1.570796251296997j)
TensorFlow asinh result: (19.064302444458008-1.570796251296997j)
PyTorch asinh result: (19.064302444458008-1.570796251296997j)
JAX asinh result: (19.064302444458008-1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (-0.27558019757270813+5103496.5j)
NumPy asinh result: (-16.13858413696289+1.570796251296997j)
TensorFlow asinh result: (-16.13858413696289+1.570796251296997j)
PyTorch asinh result: (-16.13858413696289+1.570796251296997j)
JAX asinh result: (-16.13858413696289+1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (-4.466612339019775+19856552j)
NumPy asinh result: (-17.4971923828125+1.5707961320877075j)
TensorFlow asinh result: (-17.4971923828125+1.5707961320877075j)
PyTorch asinh result: (-17.4971923828125+1.5707961320877075j)
JAX asinh result: (-17.4971923828125+1.5707961320877075j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (-5.850226879119873-70153976j)
NumPy asinh result: (-18.759349822998047-1.570796251296997j)
TensorFlow asinh result: (-18.759349822998047-1.570796251296997j)
PyTorch asinh result: (-18.759349822998047-1.570796251296997j)
JAX asinh result: (-18.759349822998047-1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (-7.956832408905029-69043872j)
NumPy asinh result: (-18.74340057373047-1.570796251296997j)
TensorFlow asinh result: (-18.74340057373047-1.570796251296997j)
PyTorch asinh result: (-18.74340057373047-1.570796251296997j)
JAX asinh result: (-18.74340057373047-1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (-6.05452823638916+78301960j)
NumPy asinh result: (-18.869230270385742+1.570796251296997j)
TensorFlow asinh result: (-18.869230270385742+1.570796251296997j)
PyTorch asinh result: (-18.869230270385742+1.570796251296997j)
JAX asinh result: (-18.869230270385742+1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (0.23627789318561554+66944604j)
NumPy asinh result: (18.7125244140625+1.5707963705062866j)
TensorFlow asinh result: (18.7125244140625+1.5707963705062866j)
PyTorch asinh result: (18.7125244140625+1.5707963705062866j)
JAX asinh result: (18.712522506713867+1.5707963705062866j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (5.8625264167785645+66359228j)
NumPy asinh result: (18.7037410736084+1.570796251296997j)
TensorFlow asinh result: (18.7037410736084+1.570796251296997j)
PyTorch asinh result: (18.7037410736084+1.570796251296997j)
JAX asinh result: (18.7037410736084+1.570796251296997j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
Generated complex number: (-0.7495654225349426-82627728j)
NumPy asinh result: (-18.923004150390625-1.5707963705062866j)
TensorFlow asinh result: (-18.923004150390625-1.5707963705062866j)
PyTorch asinh result: (-18.923004150390625-1.5707963705062866j)
JAX asinh result: (-18.923002243041992-1.5707963705062866j)
NumPy vs JAX: True
TensorFlow vs JAX: True
PyTorch vs JAX: True
==============================
```

Attaching the [gist](https://colab.sandbox.google.com/gist/rajasekharporeddy/78de9e71722528d5fb746f5d7c43a67e/-19398.ipynb) for reference.

Thank you.

**Comment by pearu** [comment]
The reproducer of this issue gives expected result:
```python
>>> import jax.numpy as jnp
>>> complex_number_jax = jnp.array([-1.0 - 101999.0j], dtype=jnp.complex64)
>>> from jax import jit
>>> asinh_fn = jit(jnp.arcsinh)
>>> result_jax = asinh_fn(complex_number_jax)
>>> print("JAX Result:", result_jax)
JAX Result: [-12.225865-1.5707866j]
```
@jakevdp @hawkinsp the issue can be closed.
